### PR TITLE
Reachability - Maven

### PR DESCRIPTION
**Pull request by meghfossa**
# Overview

This PR,
- adds maven reachability analysis
- adds reachability analysis upload mechanism
 

## Acceptance criteria
- maven reachability analysis can be computed
- maven reachability is not uploaded, if organization does not support reachability

## Testing plan

```bash
git checkout master && git pull origin && git checkout feat/reachability-in-cli-maven && make install-dev

# you can get project from: https://github.com/fossas/example-projects/tree/main/maven
# make sure you have maven installed
# make sure you have java installed (confirm by running java command)
fossa-dev analyze /path/to/maven/project -o --debug

gunzip fossa.debug.json.gz
cat fossa.debug.json | jq '.bundleReachabilityRaw' # you should see raw call graph analysis if any!
```

## References

https://fossa.atlassian.net/browse/ANE-1412

## Checklist

- [x] I added tests for this PR's change (or explained in the PR description why tests don't make sense).
- [x] If this PR introduced a user-visible change, I added documentation into `docs/`.
- [x] If this PR added docs, I added links as appropriate to the user manual's ToC in `docs/README.ms` and gave consideration to how discoverable or not my documentation is.
- [x] If this change is externally visible, I updated `Changelog.md`. If this PR did not mark a release, I added my changes into an `# Unreleased` section at the top.
- [x] If I made changes to `.fossa.yml` or `fossa-deps.{json.yml}`, I updated `docs/references/files/*.schema.json` AND I have updated example files used by `fossa init` command. You may also need to update these if you have added/removed new dependency type (e.g. `pip`) or analysis target type (e.g. `poetry`).
- [x] If I made changes to a subcommand's options, I updated `docs/references/subcommands/<subcommand>.md`.

### CODE REVIEW

File: src/App/Fossa/Analyze/Debug.hs
Comment (meghfossa) [inline src/App/Fossa/Analyze/Debug.hs:102]: For easier debugging in future

File: src/App/Fossa/Analyze/Debug.hs
Comment (meghfossa) [inline src/App/Fossa/Analyze/Debug.hs:224]: Unfortunately, there is no easy way to persist record entry in debug bundle at root level without inferring from scope. I do think we should make this easy moving forward, and more type constrained.

File: src/Strategy/Maven/Pom/PomFile.hs
Comment (meghfossa) [inline src/Strategy/Maven/Pom/PomFile.hs:194]: self-review: fix grammar

**Comment by meghfossa** [comment]
I expect to have second PR, once core side changes are merged in! This is why I've marked this PR as [Do Not Merge]

File: docs/contributing/reachability.md
Comment (meghfossa) [inline docs/contributing/reachability.md:6]: Self-review: we need to mention that java needs to be installed. Since dynamic analysis was done with maven, this is given, but we should explicitly mention it.

File: docs/contributing/reachability.md
Comment (meghfossa) [inline docs/contributing/reachability.md:8]: Self-review: mention, we need Endpoint to support reachability for reachability analysis to function

File: docs/contributing/reachability.md
Comment (JeffreyHuynh1) [inline docs/contributing/reachability.md:4]: [Question]: Where are the `3-10%` numbers coming from?

File: docs/contributing/reachability.md
Comment (JeffreyHuynh1) [inline docs/contributing/reachability.md:95]: [nit]: `You can inspect the data by running,` -> `You can inspect the data by running: `

File: src/App/Fossa/Analyze/Debug.hs
Comment (JeffreyHuynh1) [inline src/App/Fossa/Analyze/Debug.hs:25]: [nit]: Change back to `) where` for consistency with all of our other files

File: src/App/Fossa/Reachability/Maven.hs
Comment (JeffreyHuynh1) [inline src/App/Fossa/Reachability/Maven.hs:4]: [Question]: Is the `-- * for testing,` necessary?

File: src/App/Fossa/Reachability/Maven.hs
Comment (JeffreyHuynh1) [inline src/App/Fossa/Reachability/Maven.hs:61]: [nit]: Capitalization -> `"Found candidate jars: "`

File: src/App/Fossa/Reachability/Upload.hs
Comment (JeffreyHuynh1) [inline src/App/Fossa/Reachability/Upload.hs:135]: [nit]: Capitalization of log messages

File: src/App/Fossa/Reachability/Upload.hs
Comment (JeffreyHuynh1) [inline src/App/Fossa/Reachability/Upload.hs:151]: [Prefer]: `++ concatMap (\ud -> sourceDepLocator ud : sourceDepImports ud) $ buildDependencies unit`

File: src/Control/Carrier/FossaApiClient.hs
Comment (JeffreyHuynh1) [inline src/Control/Carrier/FossaApiClient.hs:66]: [Optional]: I think the `-- Reachability` comment is unnecessary as it can be implied by the function name

File: test/Reachability/MavenSpec.hs
Comment (JeffreyHuynh1) [inline test/Reachability/MavenSpec.hs:18]: Should we have tests for `isValidJar` and your other exported testing functions?

**Review by JeffreyHuynh1 — CHANGES_REQUESTED** [review]
Everything looks great for the most part! I tested the the reachability implementation locally with Megh and everything ran as expected! 

Just had some minor comments / suggestions / questions

File: src/App/Fossa/Reachability/Upload.hs
Comment (meghfossa) [reply src/App/Fossa/Reachability/Upload.hs:148]: This ran into issue prefix order in Haskell, going with what I originally had

File: src/App/Fossa/Analyze/Debug.hs
Comment (jssblck) [reply src/App/Fossa/Analyze/Debug.hs:226]: I don't understand- why not just add it like we do e.g. `bundleOutput`?

File: src/App/Fossa/Reachability/Upload.hs
Comment (jssblck) [inline src/App/Fossa/Reachability/Upload.hs:124]: [optional] consider "FOSSA CLI does not ..." rather than "We do not ..." here and in the other log message

**Review by jssblck — APPROVED** [review]

File: src/App/Fossa/Analyze/Debug.hs
Comment (meghfossa) [reply src/App/Fossa/Analyze/Debug.hs:228]: If we want to pipe it similar to `bundleOutput`, we would have to have Analyze's return value, include record of interest. I found this easier, instead of wiring it across nested function calls.

**Review by JeffreyHuynh1 — APPROVED** [review]
LGTM!

**Pull request merged**
